File: frontend/src/domain/valuesobject/Email.ts
export class Email {
  private constructor(private email: string) {}

  public static create(email: string): Email {
    this.validate(email);
    return new Email(email.trim());
  }
  private static validate(email: string): void {
    if (!email || typeof email !== 'string' || email.trim().length === 0) {
      throw new Error('senha é obrigatorio');
    }
  }
  public getValue(): string {
    return this.email;
  }
}

File: backend/src/domain/valuesobject/Email.go
package valuesobject

import (
	"errors"
	"regexp"
	"strings"
)

type Email struct {
	email string
}

func NewEmail(email string) (*Email, error) {
	trimmed := strings.TrimSpace(email)
	if err := validateEmail(trimmed); err != nil {
		return nil, err
	}
	return &Email{email: trimmed}, nil
}

func (e *Email) GetValue() string {
	return e.email
}

func validateEmail(email string) error {
	if len(email) == 0 {
		return errors.New("email é obrigatorio")
	}

	isValid := regexp.MustCompile(`^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$`)

	if !isValid.MatchString(email) {
		return errors.New("e-mail inválido. Use um formato como exemplo@dominio.com")
	}
	
	return nil
}
